system/lib/wasmfs/syscalls.cpp

```diff
@@ -1389,6 +1391,82 @@ int __syscall_fcntl64(int fd, int cmd, ...) {
   }
 }
 
+intptr_t _mmap_js(intptr_t addr,
+                  size_t length,
+                  int prot,
+                  int flags,
+                  int fd,
+                  size_t offset,
+                  int* allocated) {
+  if (addr != 0) {
+    // We don't currently support location hints for the address of the mapping.
+    return -EINVAL;
+  }
+
+  auto openFile = wasmFS.getFileTable().locked().getEntry(fd);
+  if (!openFile) {
+    return -EBADF;
+  }
+
+  std::shared_ptr<DataFile> file;
+
+  // Keep the open file info locked only for as long as we need that.
+  {
+    auto lockedOpenFile = openFile->locked();
+
+    // Check permissions. We always need read permissions, since we need to read
+    // the data in the file to map it.
+    if ((lockedOpenFile.getFlags() & O_ACCMODE) == O_WRONLY) {
+      return -EACCES;
+    }
+
+    // According to the POSIX spec it is possible to write to a file opened in
+    // read-only mode with MAP_PRIVATE flag, as all modifications will be
+    // visible only in the memory of the current process.
+    if ((prot & PROT_WRITE) != 0 && (flags & MAP_PRIVATE) == 0 &&
+        (lockedOpenFile.getFlags() & O_ACCMODE) != O_RDWR) {
+      return -EACCES;
+    }
+
+    file = lockedOpenFile.getFile()->dynCast<DataFile>();
+  }
+
+  if (!file) {
+    return -ENODEV;
+  }
+
+  // TODO: handle MAP_PRIVATE in an optimal way from here
+  // TODO: optimize here, do not always allocate + copy into a new region
+
+  // Align to a wasm page size, as we expect in the future to get wasm
+  // primitives to do this work, and those would presumably be aligned to a page
+  // size. Aligning now avoids confusion later.
+  uint8_t* ptr = (uint8_t*)memalign(WASM_PAGE_SIZE, length);
+  if (!ptr) {
+    return -ENOMEM;
+  }
+  *allocated = true;
+
+  auto written = file->locked().read(ptr, length, offset);
+  if (written < 0) {
+    // The read failed. Report the error, but first free the allocation.
+    free(ptr);
+    return written;
+  }
+
+  // From here on, we have succeeded, and can mark the allocation as having
+  // occurred (which means that the caller has the responsibility to free it).
+  *allocated = true;
+
+  // The read must be of a valid amount, or we have had an internal logic error.
+  assert(written <= length);
+
+  // mmap clears any extra bytes after the data itself.
+  memset(ptr + written, 0, length - written);
+
+  return (intptr_t)ptr;
```

Comment:
I wonder if we could just implement this in terms of `read` .. then it could share the implementation with `system/lib/standalone/standalone.c` which currently has a TODO to implement this.

Perhaps could call this `fake_mmap` and call it both from here and from `system/lib/standalone/standalone.c`  in the body of `_mmap_js`?

Reply:
> I wonder if we could just implement this in terms of `read`

I'm not sure what you're proposing here. This is written using `read` atm:
```cpp
  auto written = file->locked().read(ptr, length, offset);
```
?

Reply:
I mean we could potentially push the fake mmap into userspace and avoid a specific syscall for it. 

This would avoid the syscall the reduce the duplication.   Since we currently implement fake-mmap-based on read in several different places (here, system/lib/standalone/standalone.c and in the old FS).  There is nothing wasmfs specific about this code IIUC.

Reply:
(That would start to add WasmFS-specific stuff like an mmap hook on backends etc.)

Reply:
E.g. we could do what MEMFS does today, where if a file is backed by linear memory then a non-private mmap can literally get a pointer to that memory and avoid allocation+copy. Only the Memory backend could do that for now (but also future caching backends?). And if/when wasm adds magic for mmap-like things to the spec we could use those as well for even more things.

Reply:
oh wow that's a scary thought. Consider what would happen if someone resizes the file while it is mapped like that... I don't plan on implementing caches with contiguous whole-file memory storage, either, so this would literally only work for the in-memory files.

Reply:
Do we absolutely need to support that MEMFS-style mapping? If not, I would prefer to implement mmap in userspace like @sbc100 suggests and deprecate that mapping behavior.

Reply:
Yes, I think we should do this (at least until wasm gives is something lime mmap in the distant future).

Its hard to imagine applications is relying in MEMFS + mmap being zero-copy ..

Reply:
BTW I don't think we need to block this change on any of this